api/src/main/java/org/eclipse/microprofile/reactive/messaging/Emitter.java

```diff
@@ -0,0 +1,86 @@
+ * Copyright (c) 2018-2019 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * You may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.eclipse.microprofile.reactive.messaging;
+
+import java.util.concurrent.CompletionStage;
+
+/**
+ * Interface used to feed a channel from an <em>imperative</em> piece of code.
+ * <p>
+ * Instances are injected using:
+ *
+ * <pre>
+ * &#64;Inject
+ * &#64;Channel("my-channel")
+ * Emitter&lt;String&gt; emitter;
+ * </pre>
+ * <p>
+ * You can use an injected emitter to send either payload or
+ * {@link org.eclipse.microprofile.reactive.messaging.Message Messages}.
+ * <p>
+ * The name of the channel (given in the {@link Channel Channel annotation})
+ * indicates which channel is fed. It must match the name used in a method using
+ * {@link org.eclipse.microprofile.reactive.messaging.Incoming @Incoming} or an
+ * outgoing channel configured in the application configuration.
+ *
+ * @param <T> type of payload 
+ */
+public interface Emitter<T> {
+
+    /**
+     * Sends a payload to the channel.
+     *
+     * @param msg the <em>thing</em> to send, must not be {@code null}
+     * @return the {@code CompletionStage}, which will be completed as sending the payload alone does not provide a callback
+     *         mechanism.
+     * @throws IllegalStateException if the channel has been cancelled or terminated.
+     */
+    CompletionStage<Void> send(T msg);
+
+    /**
+     * Sends a message to the channel.
+     * @param <M> the <em>Message</em> type
+     * @param msg the <em>Message</em> to send, must not be {@code null}
+     * @throws IllegalStateException if the channel has been cancelled or terminated.
+     */
+    <M extends Message<? extends T>> void send(M msg);
+
+    /**
+     * Sends the completion event to the channel indicating that no other events will be sent afterward. 
+     */
+    void complete();
+
+    /**
+     * Sends a failure event to the channel. No more events will be sent afterward.
+     *
+     * @param e the exception, must not be {@code null}
+     */
+    void error(Exception e);
+
+    /**
+     * @return {@code true} if the emitter has been terminated or the subscription cancelled.
+     */
+    boolean isCancelled();
+
+    /**
+     * @return {@code true} if a subscriber accepts messages, {@code false} otherwise.
+     */
+    boolean isRequested();
```

Comment:
I'm still not totally clear about when this method will return `true` and when it will return `false`.

Let me check if I have this right:

This method will return `true` if:
 - A subscriber has subscribed to the publisher which backs this `Emitter`
 - That subscriber has outstanding demand (i.e. it has indicated more demand using `request()` than it has received messages using `onNext()`.

Otherwise it will return `false`.

For example, it would return `false` in these situations:
 - Nothing has connected to the channel
 - Something is connected to the channel, but it hasn't requested anything yet (maybe because it's a processor and no-one has subscribed to it yet)
 - Something is connected to the channel, and has indicated demand, but all that demand has been met by messages sent through the `Emitter`.
  - This is still true, even if the emitter is annotated with `@OnOverflow(UNBOUNDED_BUFFER)` and so the user can still call `send()` without causing an error.

Reply:
Seems like I have understood it wrong, i was expecting it returning true if ever any request signal was intercepted no matter what is actual demand. If it really returns true only if demand is higher than 0, it seems to be open invitation for a race condition right?

Reply:
This method is mostly for monitoring. Yes, the returned value may be not actual anymore, but if you get "false" for a _long time_, something is probably wrong (nothing connected along the line, consumers not requesting anymore...).

Reply:
The method `isRequested()` returns true if the subscriber accepts the message. I am confused why it is not clear.

Reply:
It could be renamed to `hasRequests` or `hasPendingRequests`. 

Looking at Reactor and RX:
Reactor uses `requestedFromDownstream` which returns a `long` (so this value is just informative). RX uses `requested` which returns also the number of pending requests.

We could do the same.

Reply:
@cescoffier I am not sure how useful to get the remaining number of requests to be processed. If we just need to know there are pending requests, we can just update the method to be `boolean hasRequests()`, but it is different from the previous method. The other thought is to remove the whole method in the first draft and then add one after we are certain with the usage.

Reply:
As I mentioned, the usage is mostly monitoring.

Reply:
ok. Are you happy with the following:
 `boolean hasRequests()` - returns true if one or more subscribers request messages from the corresponding channel where the emitter connects to.

Reply:
@cescoffier see the above suggestion to see whether you are ok with it.

Reply:
I'm ok with `hasRequests`

Reply:
ok. done